### PR TITLE
Add parent selection button to libraries and dilutions

### PR DESCRIPTION
**Pull request by apmasell**
JIRA: GLT-2206

### CODE REVIEW

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/ParentFinder.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    public final HttpEntity<byte[]> handle(ObjectMapper mapper, Stream<M> items) throws JsonProcessingException {
+      List<String> errors = new ArrayList<>();
+      List<D> parents = items.map(item -> find(item, errors::add))//
+          .filter(Objects::nonNull)//
+          .collect(Collectors.groupingBy(Identifiable::getId)).values().stream()//
+          .map(l -> l.get(0))//
+          .map(this::asDto)//
+          .collect(Collectors.toList());
+      if (errors.size() > 0) {
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/ParentFinder.java]: MINOR Use isEmpty() to check whether the collection is empty or not. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/ParentFinder.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import uk.ac.bbsrc.tgac.miso.core.data.DetailedSample;
+import uk.ac.bbsrc.tgac.miso.core.data.Identifiable;
+import uk.ac.bbsrc.tgac.miso.core.data.Sample;
+import uk.ac.bbsrc.tgac.miso.core.util.LimsUtils;
+import uk.ac.bbsrc.tgac.miso.core.util.WhineyFunction;
+import uk.ac.bbsrc.tgac.miso.dto.Dtos;
+import uk.ac.bbsrc.tgac.miso.dto.SampleDto;
+
+public abstract class ParentFinder<M extends Identifiable> {
+  public static abstract class ParentAdapter<M extends Identifiable, P extends Identifiable, D> {
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/ParentFinder.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AModifiersOrderCheck)

**Comment by sqwci** [comment]
SonarQube analysis reported 8 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [SampleRestController.java#L266](https://github.com/TGAC/miso-lims/blob/78565a82c98502e4094b7da273cd9601c35766fb/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/SampleRestController.java#L266): Remove usage of generic wildcard type. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1452)
1. ![MAJOR][MAJOR] [SampleRestController.java#L178](https://github.com/TGAC/miso-lims/blob/78565a82c98502e4094b7da273cd9601c35766fb/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/SampleRestController.java#L178): Replace the usage of the "instanceof" operator by a catch block. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1193)
1. ![MAJOR][MAJOR] [hot.js#L51](https://github.com/TGAC/miso-lims/blob/78565a82c98502e4094b7da273cd9601c35766fb/miso-web/src/main/webapp/scripts/hot.js#L51): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L228](https://github.com/TGAC/miso-lims/blob/78565a82c98502e4094b7da273cd9601c35766fb/miso-web/src/main/webapp/scripts/hot.js#L228): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MINOR][MINOR] [hot_sample.js#L599](https://github.com/TGAC/miso-lims/blob/78565a82c98502e4094b7da273cd9601c35766fb/miso-web/src/main/webapp/scripts/hot_sample.js#L599): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)
1. ![MINOR][MINOR] [hot_sample.js#L605](https://github.com/TGAC/miso-lims/blob/78565a82c98502e4094b7da273cd9601c35766fb/miso-web/src/main/webapp/scripts/hot_sample.js#L605): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sqwci** [comment]
SonarQube analysis reported 8 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [SampleRestController.java#L266](https://github.com/TGAC/miso-lims/blob/51fb53ebb889c4c3347fa9730ba63eedc5204ed1/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/SampleRestController.java#L266): Remove usage of generic wildcard type. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1452)
1. ![MAJOR][MAJOR] [SampleRestController.java#L178](https://github.com/TGAC/miso-lims/blob/51fb53ebb889c4c3347fa9730ba63eedc5204ed1/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/SampleRestController.java#L178): Replace the usage of the "instanceof" operator by a catch block. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1193)
1. ![MAJOR][MAJOR] [hot.js#L51](https://github.com/TGAC/miso-lims/blob/51fb53ebb889c4c3347fa9730ba63eedc5204ed1/miso-web/src/main/webapp/scripts/hot.js#L51): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L228](https://github.com/TGAC/miso-lims/blob/51fb53ebb889c4c3347fa9730ba63eedc5204ed1/miso-web/src/main/webapp/scripts/hot.js#L228): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MINOR][MINOR] [hot_sample.js#L599](https://github.com/TGAC/miso-lims/blob/51fb53ebb889c4c3347fa9730ba63eedc5204ed1/miso-web/src/main/webapp/scripts/hot_sample.js#L599): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)
1. ![MINOR][MINOR] [hot_sample.js#L605](https://github.com/TGAC/miso-lims/blob/51fb53ebb889c4c3347fa9730ba63eedc5204ed1/miso-web/src/main/webapp/scripts/hot_sample.js#L605): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/ParentFinder.java (hunk 2 above)
Comment (apmasell) [reply miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/ParentFinder.java]: Thank you Sonar, for not telling me the correct order here or in the detailed information nor linking me to the correct section of the JLS.

**Comment by apmasell** [comment]
Fixed. I had the wrong URL slug. :grimacing:

**Pull request merged**